File: src/Strength/SolidNodeList.cc
//---------------------------------Spheral++----------------------------------//
// SolidNodeList -- A form of the SPH NodeList appropriate for use with 
// solid materials.
//
// Created by JMO, Tue Sep 7 22:44:37 2004
//----------------------------------------------------------------------------//
#include "SolidNodeList.hh"
#include "Field/Field.hh"
#include "Field/FieldList.hh"
#include "Material/EquationOfState.hh"
#include "Kernel/TableKernel.hh"
#include "FileIO/FileIO.hh"
#include "Hydro/HydroFieldNames.hh"
#include "Strength/SolidFieldNames.hh"
#include "DataBase/IncrementState.hh"
#include "SolidMaterial/StrengthModel.hh"

#include "Utilities/DBC.hh"

namespace Spheral {
namespace SolidMaterial {

using namespace std;

using Material::EquationOfState;
using FieldSpace::Field;
using FieldSpace::FieldList;
using FileIOSpace::FileIO;
using NodeSpace::FluidNodeList;

//------------------------------------------------------------------------------
// Construct with the given EOS object, along with optional numInternal nodes,
// numGhost nodes, and name.
//------------------------------------------------------------------------------
template<typename Dimension>
SolidNodeList<Dimension>::
SolidNodeList(string name,
              EquationOfState<Dimension>& eos,
              SolidMaterial::StrengthModel<Dimension>& strength,
              const int numInternal,
              const int numGhost,
              const Scalar hmin,
              const Scalar hmax,
              const Scalar hminratio,
              const Scalar nPerh,
              const int maxNumNeighbors,
              const Scalar rhoMin,
              const Scalar rhoMax):
  FluidNodeList<Dimension>(name, 
                           eos,
                           numInternal, 
                           numGhost,
                           hmin,
                           hmax,
                           hminratio,
                           nPerh,
                           maxNumNeighbors,
                           rhoMin,
                           rhoMax),
  mDeviatoricStress(SolidFieldNames::deviatoricStress, *this),
  mPlasticStrain(SolidFieldNames::plasticStrain, *this),
  mPlasticStrainRate(SolidFieldNames::plasticStrainRate, *this),
  mDamage(SolidFieldNames::tensorDamage, *this),
  mEffectiveDamage(SolidFieldNames::effectiveTensorDamage, *this),
  mDamageGradient(SolidFieldNames::damageGradient, *this),
  mFragmentIDs(SolidFieldNames::fragmentIDs, *this),
  mStrength(strength) {
}

//------------------------------------------------------------------------------
// Destructor.
//------------------------------------------------------------------------------
template<typename Dimension>
SolidNodeList<Dimension>::
~SolidNodeList() {
}

//------------------------------------------------------------------------------
// Calculate and return the sound speed field.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
soundSpeed(Field<Dimension, typename Dimension::Scalar>& field) const {

  // Get the straight EOS look up from the base class.
  FluidNodeList<Dimension>::soundSpeed(field);

  // Augment the sound speed with the strength model.
  const Field<Dimension, Scalar>& rho = this->massDensity();
  const Field<Dimension, Scalar>& u = this->specificThermalEnergy();
  Field<Dimension, Scalar> P(HydroFieldNames::pressure, *this);
  this->pressure(P);
  for (int i = 0; i != this->numInternalNodes(); ++i) {
    field(i) = mStrength.soundSpeed(rho(i), u(i), P(i), field(i));
  }
}

//------------------------------------------------------------------------------
// Calculate and return the bulk modulus.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
bulkModulus(Field<Dimension, typename Dimension::Scalar>& field) const {
  const Field<Dimension, Scalar>& rho = this->massDensity();
  const Field<Dimension, Scalar>& u = this->specificThermalEnergy();
  const EquationOfState<Dimension>& eos = this->equationOfState();
  this->equationOfState().setBulkModulus(field, rho, u);
}

//------------------------------------------------------------------------------
// Calculate and return the shear modulus.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
shearModulus(Field<Dimension, typename Dimension::Scalar>& field) const {
  const Field<Dimension, Scalar>& rho = this->massDensity();
  const Field<Dimension, Scalar>& u = this->specificThermalEnergy();
  Field<Dimension, Scalar> P(HydroFieldNames::pressure, *this);
  this->pressure(P);
  mStrength.shearModulus(field, rho, u, P);
}

//------------------------------------------------------------------------------
// Calculate and return the yield strength.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
yieldStrength(Field<Dimension, typename Dimension::Scalar>& field) const {
  const Field<Dimension, Scalar>& rho = this->massDensity();
  const Field<Dimension, Scalar>& u = this->specificThermalEnergy();
  Field<Dimension, Scalar> P(HydroFieldNames::pressure, *this);
  this->pressure(P);
  const Field<Dimension, SymTensor>& D = this->effectiveDamage();
  mStrength.yieldStrength(field, rho, u, P, mPlasticStrain, mPlasticStrainRate);
}

//------------------------------------------------------------------------------
// Dump the current state of the NodeList to the given file.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
dumpState(FileIO& file, const string& pathName) const {

  // Dump the ancestor class.
  FluidNodeList<Dimension>::dumpState(file, pathName);

  file.write(mDeviatoricStress, pathName + "/" + mDeviatoricStress.name());
  file.write(mPlasticStrain, pathName + "/" + mPlasticStrain.name());
  file.write(mPlasticStrainRate, pathName + "/" + mPlasticStrainRate.name());
  file.write(mDamage, pathName + "/" + mDamage.name());
  file.write(mEffectiveDamage, pathName + "/" + mEffectiveDamage.name());
  file.write(mDamageGradient, pathName + "/" + mDamageGradient.name());
  file.write(mFragmentIDs, pathName + "/" + mFragmentIDs.name());
}

//------------------------------------------------------------------------------
// Restore the state of the NodeList from the given file.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
restoreState(const FileIO& file, const string& pathName) {

  // Restore the ancestor class.
  FluidNodeList<Dimension>::restoreState(file, pathName);

  file.read(mDeviatoricStress, pathName + "/" + mDeviatoricStress.name());
  file.read(mPlasticStrain, pathName + "/" + mPlasticStrain.name());
  file.read(mPlasticStrainRate, pathName + "/" + mPlasticStrainRate.name());
  file.read(mDamage, pathName + "/" + mDamage.name());
  file.read(mEffectiveDamage, pathName + "/" + mEffectiveDamage.name());
  file.read(mDamageGradient, pathName + "/" + mDamageGradient.name());
  file.read(mFragmentIDs, pathName + "/" + mFragmentIDs.name());
}

}
}

File: src/PYB11/NodeList/SolidNodeList.py
from PYB11Generator import *
from FluidNodeList import FluidNodeList
from RestartMethods import *

#-------------------------------------------------------------------------------
# SolidNodeList template
#-------------------------------------------------------------------------------
@PYB11template("Dimension")
@PYB11module("SpheralNodeList")
@PYB11dynamic_attr
class SolidNodeList(FluidNodeList):
    "Spheral SolidNodeList base class in %(Dimension)s, i.e.,  the NodeList for solid dynamics."

    PYB11typedefs = """
    typedef typename %(Dimension)s::Scalar Scalar;
    typedef typename %(Dimension)s::Vector Vector;
    typedef typename %(Dimension)s::Tensor Tensor;
    typedef typename %(Dimension)s::SymTensor SymTensor;
    typedef Field<%(Dimension)s, int> IntField;
    typedef Field<%(Dimension)s, Scalar> ScalarField;
    typedef Field<%(Dimension)s, Vector> VectorField;
    typedef Field<%(Dimension)s, Tensor> TensorField;
    typedef Field<%(Dimension)s, SymTensor> SymTensorField;
"""

    def pyinit(self,
               name = "std::string",
               eos = "EquationOfState<%(Dimension)s>&",
               strength = "StrengthModel<%(Dimension)s>&",
               numInternal = ("int", "0"),
               numGhost = ("int", "0"),
               hmin = ("double", "1e-20"),
               hmax = ("double", "1e20"),
               hminratio = ("double", "0.1"),
               nPerh = ("double", "2.01"),
               maxNumNeighbors = ("int", "500"),
               rhoMin = ("double", "1e-10"),
               rhoMax = ("double", "1e100")):
        "Constructor for a SolidNodeList class."
        return

    @PYB11virtual
    @PYB11const
    def soundSpeed(self, result="ScalarField&"):
        "Compute the current sound speed, storing the result in the argument ScalarField."
        return "void"

    @PYB11virtual
    @PYB11const
    def bulkModulus(self, result="ScalarField&"):
        "Compute the current bulk modulus, storing the result in the argument ScalarField."
        return "void"

    @PYB11virtual
    @PYB11const
    def yieldStrength(self, result="ScalarField&"):
        "Compute the current yield strength, storing the result in the argument ScalarField."
        return "void"

    @PYB11virtual
    @PYB11const
    def YoungsModulus(self,
                      result="ScalarField&",
                      K = "const ScalarField&",
                      mu = "const ScalarField&"):
        "Compute the current Youngs modulus, storing the result in the argument result, given the bulk modulus (K) and shear modulus (mu)."
        return "void"

    @PYB11virtual
    @PYB11const
    def longitudinalSoundSpeed(self,
                               result="ScalarField&",
                               rho = "const ScalarField&",
                               K = "const ScalarField&",
                               mu = "const ScalarField&"):
        """Compute the current longitudinal sound speed, storing the result in the argument result,
given the mass density, bulk modulus, and shear modulus (rho, K, mu)"""
        return "void"

    @PYB11const
    @PYB11returnpolicy("reference_internal")
    def deviatoricStress(self):
        "The mass density field"
        return "const SymTensorField&"

    @PYB11const
    @PYB11returnpolicy("reference_internal")
    def plasticStrain(self):
        "The plastic strain field"
        return "const ScalarField&"

    @PYB11const
    @PYB11returnpolicy("reference_internal")
    def plasticStrainRate(self):
        "The plastic strain rate field"
        return "const ScalarField&"

    @PYB11const
    @PYB11returnpolicy("reference_internal")
    def damage(self):
        "The damage field"
        return "const SymTensorField&"

    @PYB11const
    @PYB11returnpolicy("reference_internal")
    def fragmentIDs(self):
        "The fragment IDs field"
        return "const IntField&"

    @PYB11const
    @PYB11returnpolicy("reference_internal")
    def particleTypes(self):
        "The particle type field"
        return "const IntField&"

    @PYB11const
    @PYB11returnpolicy("reference_internal")
    def strengthModel(self):
        "Return the strength model object this SolidNodeList is associated with."
        return "const StrengthModel<%(Dimension)s>&"

    # Comparison
    def __eq__(self):
        "Equivalence test with another SolidNodeList"

    def __ne__(self):
        "Inequivalence test with another SolidNodeList"

    @PYB11implementation("[](const SolidNodeList<%(Dimension)s>& self) -> std::uintptr_t { return reinterpret_cast<std::uintptr_t>(&self); }")
    def __hash__(self):
        "Make SolidNodeList objects hashable for Python"
        return "std::uintptr_t"

#-------------------------------------------------------------------------------
# Inject the restart methods
#-------------------------------------------------------------------------------
PYB11inject(RestartMethods, SolidNodeList)
